lib/types.js

```diff
@@ -678,9 +685,36 @@ Dp.finalize = function() {
 
         // A linearization of the inheritance hierarchy.
         populateSupertypeList(this.typeName, this.supertypeList);
+
+        var exp = Type.def("Expression");
+        // Calling `exp.isSupertypeOf(...)` will throw if `exp` has not been finalized.
+        // We finalize the entire supertype chain above, so if `exp` is not yet finalized
+        // we can safely assume that this is not an Expression.
+        if (this.buildable && exp.finalized && exp.isSupertypeOf(this)) {
```

Comment:
Since `this.supertypeList` has just been populated, this test can be simpler:
```js
if (this.buildable && this.supertypeList.lastIndexOf("Expression") >= 0) {
```